ql/src/java/org/apache/hadoop/hive/ql/parse/DeleteSemanticAnalyzer.java

```diff
@@ -0,0 +1,165 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.parse;
+
+import org.apache.hadoop.hive.common.TableName;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.Context;
+import org.apache.hadoop.hive.ql.QueryState;
+import org.apache.hadoop.hive.ql.ddl.DDLWork;
+import org.apache.hadoop.hive.ql.ddl.table.execute.AlterTableExecuteDesc;
+import org.apache.hadoop.hive.ql.exec.TableScanOperator;
+import org.apache.hadoop.hive.ql.exec.TaskFactory;
+import org.apache.hadoop.hive.ql.io.sarg.ConvertAstToSearchArg;
+import org.apache.hadoop.hive.ql.io.sarg.SearchArgument;
+import org.apache.hadoop.hive.ql.lib.Node;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.hadoop.hive.ql.parse.rewrite.DeleteStatement;
+import org.apache.hadoop.hive.ql.parse.rewrite.Rewriter;
+import org.apache.hadoop.hive.ql.parse.rewrite.RewriterFactory;
+import org.apache.hadoop.hive.ql.plan.ExprNodeGenericFuncDesc;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class DeleteSemanticAnalyzer extends RewriteSemanticAnalyzer {
+
+  private final RewriterFactory<DeleteStatement> rewriterFactory;
+
+  public DeleteSemanticAnalyzer(QueryState queryState, RewriterFactory<DeleteStatement> rewriterFactory)
+      throws SemanticException {
+    super(queryState);
+    this.rewriterFactory = rewriterFactory;
+  }
+
+  @Override
+  protected ASTNode getTargetTableNode(ASTNode tree) {
+    // The first child should be the table we are updating / deleting from
+    ASTNode tabName = (ASTNode) tree.getChild(0);
+    assert tabName.getToken().getType() == HiveParser.TOK_TABNAME :
+        "Expected tablename as first child of " + Context.Operation.DELETE + " but found " + tabName.getName();
+    return tabName;
+  }
+
+  @Override
+  protected void analyze(ASTNode tree, Table table, ASTNode tableName) throws SemanticException {
+    List<? extends Node> children = tree.getChildren();
+
+    ASTNode where = null;
+    if (children.size() > 1) {
+      where = (ASTNode) children.get(1);
+      assert where.getToken().getType() == HiveParser.TOK_WHERE :
+          "Expected where clause, but found " + where.getName();
+    }
+
+    boolean shouldTruncate = HiveConf.getBoolVar(conf, HiveConf.ConfVars.HIVE_OPTIMIZE_REPLACE_DELETE_WITH_TRUNCATE)
+        && where == null;
+    if (shouldTruncate) {
```

Comment:
could we create List<Predicate<>> for diff optimizations iterate and call test(), if that returns true return, otherwise go to next?

Reply:
I think MetadataUpdate doesn't fall into that pattern easily. It requires a more generic flow something like:
init -> canHandlePreParse -> parse -> canHandleAftParse -> rewrite -> canHandleAftRewrite -> compile

It can be done in a follow-up patch.

Reply:
ok